observed_generation.md

```diff
@@ -0,0 +1,74 @@
+# Generation and ObservedGeneration
+
+## Problem
+A client should be able to decide if the reconciliation of the latest `Spec`
+update has been finished by looking at the `Status` field of the CR. The
+`Status.Conditions`, and especially the `Ready` condition in that list, can be
+used to determine the result of the last reconciliation run. However the
+conditions alone cannot tell which version of the `Spec` was used for the last
+reconciliation, they cannot indicate if there is a newer `Spec` update that has
+not seen by the controller and therefore hasn't reconciled yet.
+
+
+## Solution
+K8s maintains a `Generation` field for each CR. When the `Spec` field of the
+CR is updated the `Generation` of the CR is automatically incremented.
+
+To allow the client to decide if the `Status` field talks about the latest
+version of the `Spec` the `Status` field should tell which version, generation,
+the `Status` is referring to. This is implemented by the
+`Status.ObservedGeneration` field. This field contains the `Generation` value
+of the `Spec` the controller last seen and therefore the `Spec` version the
+`Status` is referring to.
+
+## Usage
+If a client updates the `Spec` field of a CR and wants to decide when such
+`Spec` changes are applied to the running system, then the client needs to look
+at two fields in the CR `Status`:
+* If the `Status.ObservedGeneration == Generation` then the client can be sure
+that the controller seen the last `Spec` version
+* If the `Ready` condition's State is `True` in the `Status.Conditions` list
+then the  client can be sure that the last reconciliation finished
+successfully.
+
+```golang
+if subCR.Generation == subCR.Status.ObservedGeneration &&
+	subCR.Status.Conditions.IsTrue(condition.ReadyCondition) {
+        // subCR is successfully reconciled the latest Spec update
+}
+```
+
+## Implementation
+The `Generation` field is implemented automatically.
+
+The `Status.ObservedGeneration` needs to be implemented by every controller
+individually. This field needs to be bumped to the value of the `Generation`
+field at the start of each `Reconcile` run, after the `Status.Conditions`
```

Comment:
This statement is True under the assumption that:
1. we are considering transient states a bug or just wrong
2. we are creating a strong relation between `observedGeneration` and the readiness

We can still decouple the two things, but I suspect it's a matter of having an agreement on the pattern.

Reply:
> we are considering transient states a bug or just wrong

if the transient Ready state can cause the client to think that the last spec generation is reconciled while it isn't, then I consider that as a bug. However if along with Ready state there is an observedGeneration in the status that show what this readiness is actually about an older generation than the latest Spec generation then the client has enough input to ignore the Ready state and then it is not a bug as the Status is self consistent.

> we are creating a strong relation between observedGeneration and the readiness

From the subCR controller perspective the ReadyCondition (and all other Status fields) are calculated for a given Spec generation, and this generation value is stored in the observedGeneration. I don't see how this coupling can be avoided and still present a consistent Status field.

Reply:
Yeah, I think we're on the same page for this particular point. The client will check the && condition between the Readiness and the ObservedGeneration, which makes sense when we state that: "A reconciler looks for the readiness associated to the last updated CR".
To make this happen, we need the `observedGeneration` field in the subCRs and it can be bumped up as long as a reconciling is triggered.

Reply:
ah, yes, transient state can be considered as a bug, I agree on that.

Reply:
OK sounds like we are on the same page